File: storage/db-core/src/main/java/com/coffee/gu/payment/PaymentEntity.java
package com.coffee.gu.payment;

import com.coffee.gu.Principal;
import com.coffee.gu.enums.PaymentMethod;
import com.coffee.gu.enums.PaymentState;
import com.coffee.gu.enums.PrincipalType;
import com.coffee.gu.BaseEntity;
import jakarta.persistence.*;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

@Table(
        name = "payment",
        indexes = {
                @Index(name = "udx_order_key", columnList = "orderKey", unique = true)
        }
)
@Entity
public class PaymentEntity extends BaseEntity {
    private String principalKey;
    @Enumerated(EnumType.STRING)
    private PrincipalType principalType;
    private String orderKey;
    private BigDecimal originalAmount;
    private Long issuedCouponId;
    private BigDecimal couponDiscount;
    private BigDecimal paidAmount;
    @Enumerated(EnumType.STRING)
    private PaymentState state;
    private String externalPaymentKey;
    @Enumerated(EnumType.STRING)
    private PaymentMethod method;
    private OffsetDateTime paidAt; // pg사, 카드사는 결제가 승인된 시점을 기준으로 수수료를 떼고 가맹점 계좌로 입금
    private String approveCode; // 카드사의 승인코드
    private int retryCount;

    public PaymentEntity() {}

    public PaymentEntity(String principalKey, PrincipalType principalType, String orderKey, BigDecimal originalAmount, Long issuedCouponId, BigDecimal couponDiscount, BigDecimal paidAmount, PaymentState state,
                         String externalPaymentKey, PaymentMethod method, OffsetDateTime paidAt, String approveCode, int retryCount) {
        this.principalKey = principalKey;
        this.principalType = principalType;
        this.orderKey = orderKey;
        this.originalAmount = originalAmount;
        this.issuedCouponId = issuedCouponId;
        this.couponDiscount = couponDiscount;
        this.paidAmount = paidAmount;
        this.state = state;
        this.externalPaymentKey = externalPaymentKey;
        this.method = method;
        this.paidAt = paidAt;
        this.approveCode = approveCode;
        this.retryCount = retryCount;
    }


    public static PaymentEntity from(Payment payment) {
        PaymentEntity entity = new PaymentEntity(
                payment.getPrincipal().getKey(),
                payment.getPrincipal().getType(),
                payment.getOrderKey(),
                payment.getOriginalAmount(),
                payment.getIssuedCouponId(),
                payment.getCouponDiscount(),
                payment.getAmount(),
                payment.getState(),
                payment.getExternalPaymentKey(),
                payment.getMethod(),
                payment.getPaidAt(),
                payment.getApproveCode(),
                payment.getRetryCount()
        );
        entity.id = payment.getId();
        return entity;
    }

    public Payment toModel() {
        return new Payment(
                this.id,
                new Principal(this.principalKey, this.principalType),
                this.orderKey,
                this.originalAmount,
                this.issuedCouponId,
                this.couponDiscount,
                this.paidAmount,
                this.state,
                this.externalPaymentKey,
                this.method,
                this.paidAt,
                this.approveCode,
                this.getCreatedAt(),
                this.retryCount
        );
    }

    public PaymentState getState() {
        return state;
    }

    public String getPrincipalKey() {
        return principalKey;
    }

    public BigDecimal getPaidAmount() {
        return paidAmount;
    }

    public OffsetDateTime getPaidAt() {
        return paidAt;
    }

    public Long getIssuedCouponId() {
        return issuedCouponId;
    }

    public PrincipalType getPrincipalType() {
        return principalType;
    }

    public String getOrderKey() {
        return orderKey;
    }

    public BigDecimal getOriginalAmount() {
        return originalAmount;
    }

    public BigDecimal getCouponDiscount() {
        return couponDiscount;
    }

    public String getExternalPaymentKey() {
        return externalPaymentKey;
    }

    public int getRetryCount() {
        return retryCount;
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/payment/PaymentEntity.kt
package com.coffee.gu.payment

import com.coffee.gu.BaseEntity
import com.coffee.gu.Principal
import com.coffee.gu.enums.PaymentMethod
import com.coffee.gu.enums.PaymentState
import com.coffee.gu.enums.PrincipalType
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.Table
import java.math.BigDecimal
import java.time.OffsetDateTime

@Table(
    name = "payment",
    indexes = [
        Index(name = "udx_order_key", columnList = "orderKey", unique = true),
    ],
)
@Entity
class PaymentEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,
    val principalKey: String,
    @Enumerated(EnumType.STRING)
    val principalType: PrincipalType,
    val orderKey: String,
    val originalAmount: BigDecimal,
    val issuedCouponId: Long? = null,
    val couponDiscount: BigDecimal? = null,
    val paidAmount: BigDecimal,
    @Enumerated(EnumType.STRING)
    var state: PaymentState,
    var externalPaymentKey: String? = null,
    @Enumerated(EnumType.STRING)
    var method: PaymentMethod? = null,
    var paidAt: OffsetDateTime? = null,
    var approveCode: String? = null,
    var retryCount: Int = 0,
) : BaseEntity() {

    fun toModel(): Payment {
        return Payment(
            id = id,
            principal = Principal(principalKey, principalType),
            orderKey = orderKey,
            originalAmount = originalAmount,
            issuedCouponId = issuedCouponId,
            couponDiscount = couponDiscount,
            amount = paidAmount,
            state = state,
            externalPaymentKey = externalPaymentKey,
            method = method,
            paidAt = paidAt,
            approveCode = approveCode,
            createdAt = createdAt,
            retryCount = retryCount,
        )
    }

    companion object {
        fun from(payment: Payment): PaymentEntity {
            return PaymentEntity(
                id = payment.id,
                principalKey = payment.principal.key,
                principalType = payment.principal.type,
                orderKey = payment.orderKey,
                originalAmount = payment.originalAmount,
                issuedCouponId = payment.issuedCouponId,
                couponDiscount = payment.couponDiscount,
                paidAmount = payment.amount,
                state = payment.state,
                externalPaymentKey = payment.externalPaymentKey,
                method = payment.method,
                paidAt = payment.paidAt,
                approveCode = payment.approveCode,
                retryCount = payment.retryCount,
            )
        }
    }
}
